### ISSUE TITLE
run player without sound

### ISSUE THREAD

**Issue by jxmorris12** [CoderLine/alphaTab#403]
# Question 

I have two questions about the AlphaTab player:
1. Is there a way to run the player (`alphaTabApi.play()`) and have the cursor scroll without sound?

2. Is there a way to dynamically update the BPM? I see that I can update the player `playbackSpeed` but I'd prefer to update the tempo of the music -- and display the new tempo on the tab -- before playing.

**Comment by Danielku15** [comment]
1. Currently this is not working out of the box, **but**. You could simply set the [`trackVolume`](https://www.alphatab.net/docs/reference/api/changetrackvolume/) to 0. Unfortunately there was also a bug in 1.0.1 which broke the masterVolume. Due to the lack of time I could not prepare a 1.0.2 bugfix release yet. But using 1.1 should still be fine from a stability perspective.  

In 1.1 I started to allow importing partial soundfonts and simply do not generate any sound for the instruments which do not have a preset loaded. I could imagine to remove some restrictions which would allow calling certain player functions without a soundfont loaded yet. The player should still operate but simply will not generate any hearable audio. Feel free to open a feature request describing your needs using the provided template. 

2. Also here there is nothing built-in that is ready to use. Changing the tempo requires a modification of the data model. There is no way to do change the tempo dynamically as it is not as simple that there would just be 1 tempo in a song which controls everything. At every (master)bar there can be a change of a tempo within the song. AlphaTab will use this information from the data model when the music sheet is rendered. For playback alphaTab generates a full midi file which is then used by soundfont midi synthesizer. Corresponding midi tempo changes are generated at the right time to ensure the right playback. This is why only an overall speedfactor during playback can easily be applied. 

To change the tempo you would need to update the `score.tempo` and the individual `score.masterBar[i].tempoAutomation` (if set) values to update the score. alphaTab [currently checks](https://github.com/CoderLine/alphaTab/blob/10e3632bcccfdc34438f8cca97720698b83a912c/src/AlphaTabApiBase.ts#L264) whether a new score was loaded or only new tracks of an existing score need to be displayed. There might be a need for a new parameter that allows to force a full reload of a score after modification. This could be bypassed currently by calling manually `api.loadMidiForScore()` (hidden) after calling `renderScore`.

Long story short, the following snippet should work. 

```js
function updateBpm(bpm) {
    const score = api.score;
    score.tempo = bpm;
    for(const mb of score.masterBars) {
        if(mb.tempoAutomation) {
            mb.tempoAutomation.value = bpm;
        }
    }
    api.loadMidiForScore(); // force regeneration of midi, unofficial API method
    api.renderTracks(at.tracks);
}
```

This will not be on-the-fly and cause a re-render and stop of the player, but it is a semi-dynamic update.

**Comment by Danielku15** [comment]
Update: as expected the latest changes related to the multiple soundfonts play well together with your need of having no soundfont at all. I just merged a small change which will allow you call play without a soundFont loaded. Instead of subscribing to the `api.playerReady` you would rather subscribe to [`api.midiLoaded`](https://www.alphatab.net/docs/reference/api/midiloaded/) to detect when only the midi has been loaded and playback can be started without sound.

**Comment by jxmorris12** [comment]
Hi @Danielku15 -- thanks so much. The BPM change worked quite well with no edits.

I've updated to v1.1.0-alpha.43 and tried initializing without passing a `soundFont` to the config for the player. That part still isn't working for me. Calling `alphaTabApi.player.play()` returns false and nothing happens. Any idea what I'm doing wrong?

**Comment by Danielku15** [comment]
Might be something special on your setup. In general it works: https://jsfiddle.net/danielku15/jxdykbmn/

**Comment by jxmorris12** [comment]
@Danielku15 it works!!

The error *was* something special about my setup: specifically, that I hadn't actually updated to 1.1.0, even though I thought that I had. Thanks for all the help, especially the jsfiddle.
